pandas-stubs/core/groupby/generic.pyi

```diff
@@ -166,7 +166,7 @@ class DataFrameGroupBy(GroupBy):
     ) -> DataFrame: ...
     def aggregate(self, arg: AggFuncTypeFrame = ..., *args, **kwargs) -> DataFrame: ...
     agg = aggregate
-    def transform(self, func, *args, **kwargs): ...
+    def transform(self, func: Callable | str, *args, **kwargs) -> DataFrame: ...
```

Comment:
Can/should we make this a `Literal`?

From https://github.com/pandas-dev/pandas/blob/main/pandas/core/groupby/base.py#L60, the transformation kernels would be the ones applicable here:

```
transformation_kernels = Literal[
    "bfill",
    "cumcount",
    "cummax",
    "cummin",
    "cumprod",
    "cumsum",
    "diff",
    "ffill",
    "fillna",
    "ngroup",
    "pct_change",
    "rank",
    "shift",
]
```

Reply:
Is there any way to do better than `Callable`? Like what about `Callable[Concatenate[pd.DataFrame, P], pd.DataFrame]`? Will `ParamSpec` work in that context? If not, can we use a `Protocol`? I know that we can't use `P.args` on `*args` and `P.kwargs` on `**kwargs`, though that would have been a very nice feature.

I'm just thinking, if we are going from an unannotated function to an annotated function, a loose annotation doesn't really do that much good versus not having an annotation at all.

Reply:
I think it would be better the ask Dr Irv about adding it.

Reply:
I'd rather be incremental in terms of PR's where we could specify the `Literal` values (there are more than just the ones you listed), as well as making the `func` argument more specific.  @gandhis1 feel free to create an issue indicating that we could improve the typing on `groupby.transform` by being more specific about the allowed argument combinations.  It also would require a lot more tests.

Focus of this PR was just to get the return types right, and do a slight improvement in the `func` argument.